engine/src/render/cascaded_shadow_maps_rasterizer/plugin.cpp

```diff
@@ -0,0 +1,314 @@
+        });
+
+    cubos.system("rasterize to cascaded shadow maps")
+        .tagged(drawToCascadedShadowMapsTag)
+        .after(createGBufferTag)
+        .call([](State& state, const Window& window, const RenderMeshPool& pool,
+                 Query<DirectionalShadowCaster&, const DirectionalLight&, const LocalToWorld&> lights,
+                 Query<Entity, const LocalToWorld&, const Camera&, Opt<const PerspectiveCamera&>,
+                       Opt<const OrthographicCamera&>, const DrawsTo&, const GBuffer&>
+                     cameras,
+                 Query<Entity, const LocalToWorld&, const RenderMesh&, const RenderVoxelGrid&> meshes) {
+            auto& rd = window->renderDevice();
+
+            for (auto [cameraEntity, cameraLocalToWorld, camera, perspectiveCamera, orthographicCamera, drawsTo,
+                       gBuffer] : cameras)
+            {
+                if (!(camera.active && (perspectiveCamera.contains() || orthographicCamera.contains())))
+                {
+                    continue;
+                }
+
+                for (auto [caster, light, localToWorld] : lights)
+                {
+                    auto& shadowMap = caster.shadowMaps.at(cameraEntity);
+                    // Check if we need to recreate the framebuffer.
+                    if (shadowMap.previousCascades != shadowMap.cascades)
+                    {
+                        // Store textures so we can check if they change in the next frame.
+                        shadowMap.previousCascades = shadowMap.cascades;
+
+                        // Create the framebuffer.
+                        FramebufferDesc desc{};
+                        desc.targetCount = 0;
+                        desc.depthStencil.setTexture2DArrayTarget(shadowMap.cascades);
+                        shadowMap.framebuffer = rd.createFramebuffer(desc);
+                    }
+
+                    // Bind the framebuffer and set the viewport.
+                    rd.setFramebuffer(shadowMap.framebuffer);
+                    rd.setViewport(0, 0, static_cast<int>(caster.size.x), static_cast<int>(caster.size.y));
+
+                    // Set the raster and depth-stencil states.
+                    rd.setRasterState(state.rasterState);
+                    rd.setBlendState(nullptr);
+                    rd.setDepthStencilState(state.depthStencilState);
+
+                    // Clear
+                    rd.clearDepth(1.0F);
+
+                    // Send the PerScene data to the GPU.
+                    PerScene perScene;
+
+                    float cameraZFar = 1000.0F;
+                    if (perspectiveCamera.contains())
+                    {
+                        cameraZFar = perspectiveCamera.value().zFar;
+                    }
+                    else if (orthographicCamera.contains())
+                    {
+                        cameraZFar = orthographicCamera.value().zFar;
+                    }
```

Comment:
Huh, I guess the `zFar` and `zNear` properties should belong to the base `Camera`. Could you move it there?
Then you would be able to abstract this over camera type.